cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1827,9 +1827,131 @@ Status ExecArrayCoalesce(KernelContext* ctx, const ExecBatch& batch, Datum* out)
   return Status::OK();
 }
 
+// Special case: implement 'coalesce' for an array and a scalar for any
+// fixed-width type (a 'fill_null' operation)
+template <typename Type>
+Status ExecArrayScalarCoalesce(KernelContext* ctx, Datum left, Datum right,
+                               int64_t length, Datum* out) {
+  ArrayData* output = out->mutable_array();
+  const int64_t out_offset = output->offset;
+  uint8_t* out_valid = output->buffers[0]->mutable_data();
+  uint8_t* out_values = output->buffers[1]->mutable_data();
+
+  const ArrayData& left_arr = *left.array();
+  const uint8_t* left_valid =
+      left_arr.MayHaveNulls() ? left_arr.buffers[0]->data() : nullptr;
+  arrow::internal::OptionalBitBlockCounter bit_counter(left_valid, left_arr.offset,
```

Comment:
Have you tried using `BitRunReader` instead?

(er, perhaps we already had this conversation? :-S)

Reply:
It was not helpful in case_when but might be useful here, I can give it a try.

Reply:
It may largely depend on the clustering of nulls/non-nulls, hence the suggestion to benchmark different null probabilites.